### ISSUE TITLE
Instancing

### ISSUE THREAD

**Issue by pc-john** [KhronosGroup/ANARI-Docs#4]
Let me start by a simple example: Say, we are creating visualization software for a city traffic and we want to render the city and all the cars inside. Each car is composed of chassis and four wheels. The chassis is in local coordinates of the car, the wheels are represented by one model instanced four times relative to the car local coordinate system. And we want to instantiate the car 1'000 times in 10 colors, so 10'000 instances in total.

In Anari, we would create two geometries - one for the chassis, another for the wheel. This is perfect and without issues.

Then, we would create 10 surfaces, assign ten different colors to them and set their geometries to the chassis. For the wheels, we would create one surface, assign the wheel as geometry and black color as material. I do not fully like such lightweight objects as surfaces that just associate geometry with material, but let's see later if there are any alternatives. Anyway, so far, we are ok.

Now, to instantiate one car, we need to create Anari::Instance that carries single transformation. We create it and assign the car transformation. We cannot associate it with the chassis surface directly, but we have to create new group and array of Surfaces and set the array's single element to the chassis surface. So, to instantiate the chassis, we had to allocate four objects - Instance, Group, Array and Surface. Luckily, the surface will be shared among 1000 objects. So, in our example, it is only three, but having unique textures for each car would result in four objects.

But what about the wheels? They cannot be placed bellow chassis instance. For this, we would need to support a kind of hierarchy of instances or multi-level instancing as you call it in #2. Currently, we would need to create Anari::Instance for each wheel and assign it composition of car and wheel transformation. I guess, your team is already aware of the problem.

However, the multi-level instancing would, unfortunately, not save the day. If all four wheels would have constant transformation against car's local coordinate system, it would work. But if we would like to change front wheel transformation during turning left or turning right and slightly rotate front wheels to follow the path of the car, it would not work. Some of 1000 cars would be turning left, some to the right and some going straight, all having different transformations of their wheels.

OpenGL Performer (famous real-time rendering library of Silicon Graphics, now a little historical) tackled the problem by introducing shared instancing and cloned instancing. With the shared instancing, all rear wheel transformations can be shared among all cars as these wheels are fixed (let's not make wheels spinning with the speed of the car, for simplicity). But all front wheels need to use cloned instancing so all cars will have their own transformation for the front wheels to allow them turning left and right.

Now, how to apply the principle to Anari? (This is not about telling that the current Anari design is not good. Rather it is about the opportunity to give arguments, why the current design is the best one and about finding arguments for it, while at the same time it is the opportunity to find other ways, possibly even better ways, that nobody thought about before.)

1. **Keep the current design** - Each car instance will result in five Anari instances - one for the chassis and four for wheels. The main disadvantages are: (a) front wheels need to receive transformation as combined matrix of particular car transformation and front wheel transformation. Many users might prefer to have two separate nodes, one for car transformation and second for front wheel transformation. (b) Rear wheels transformation needs to be updated each time car moves. Although their local transformation is static, nevertheless car movement causes the need to update their transformation.
2. **Allow for nested transformations** - This would probably result in simple change of the current Anari design, that Instance might have Instances as children. This would remove both disadvantages (a) and (b) from the previous point. The car would be composed of the single first level instance containing chassis and four second level instances, each one representing one of the wheels. The four second level instances would carry transformation in local coordinate system of the car, removing the problem of combined matrix of front wheels and removing problem of updates of rear wheels whenever the car position is updated.
3. **Introduce transformation graph** - The previous point has given us the scene graph based on coordinate transformation while parent-child relation of Instances defines the relation between parent local coordinate system and child local coordinate system. However, coupling parent-child relation with transformations might sometimes not be desirable. The parent-child relation in Anari graph gives ownership (thus existence in memory) and visibility (thus existence in rendered scene). An user might want to create model of a building with ten transformations - e.g. ten coordinate systems - one for each floor. But he might want to organize the scene by object categories, not by their transformations, e.g. by the floor on which they are placed. This would allow him to remove a single group from the scene and reinsert it later and he might hide and show again all walls of the model, or by removing one child he might hide all the furniture and later show it again by just appending again the removed child. The child would be the group holding all the objects of particular kind. This would remove coupling of the graph structure and transformations. But where to put the transformations in such case? They would have its own graph. In our example, it would have root transformation of the building and ten child-transformations for each floor. Each scene object (each Instance) would reference one of these transformations. One more additional benefits we get is the speed of transformation updates. Without transformation graph, moving all the objects on particular floor would mean changing transformation of all the objects (potentially thousands of them). With transformation graph, we just update one transformation node. Another advantage would be that having transformation graph means we can accelerate transformation computations (matrix multiplications) on GPU, removing the burden from CPU - and we might have millions of matrices in some large scenes.

Another thought was why to have all this **Instance+Group+Array+Surface** just to instantiate one Geometry. I asked myself whether instance of Geometry - is it not just Geometry+Material+Transformation? Why not having just single object holding all the three components? Something named possibly Drawable, Instance or Geode (GEOmetry noDE) and containing the references to Geometry+Material+Transformation? You might argue that we need to instantiate Volumes and Lights also, but Instance class might be made to handle all three mentioned types. You might also argue that Instance+Group+Array+Surface solution provides the ability to instantiate many Surfaces, Volumes and Lights using just one Instance. But this lacks flexibility of instancing of our four wheels in one instance of a car. It also requires all Geometries to share the same local coordinate system - basically, they must fit. All they can differ is material. If we would like this new Instance (Geometry+Material+Transformation) solution to handle more Geometries with different Materials, we might allow it to reference Array of Geometries and Materials or even better, array of Geometry+Material+Transformation. Optionally, we might call it MultiInstance. Surely, we can analyze many alternatives and look for the best fit. Or we can analyze why the current Anari design is the best one.

**Comment by jeffamstutz** [comment]
Hi,

Instead of responding to why we didn't choose X/Y/Z option, I'll instead give some rationale for what we did design. Then maybe its easier to understand the drawbacks of other ideas.

ANARI is an API to a diversity of rendering engines, all of which may use different techniques for rendering and make different trade offs for raw frame rate (frame latency), scene update rate (animations), or visual quality (materials, GI). We also know that rendering engines are diverse in the algorithms used to generate frames: pure rasterization, hybrid raster/ray tracing, or pure ray tracing.

Why does this matter?  Rasterization and ray tracing make different trade offs in what is considered "expensive" to change in a scene. Rasterization is cheap to change the geometry being drawn, but expensive to change the shader(s) controlling appearance. Conversely, ray tracing is more expensive to change the geometry being drawn (BVH rebuilds), but is generally very cheap to change materials. The (geometry + material) --> group --> instance formulation lets backend devices retain flexibility for deciding to sort geometry (ray tracing) or sort shading programs (raster). Given this arrangement of objects, any change to an object can directly express how much internal state must change to accommodate the change (rebuilt draw order or rebuild BVH(s)).

It also minimizes redundant calculations: by using groups, we can let an application directly express _exactly_ which objects share a common coordinate system -- exactly the ideal to minimize BVH overhead (both memory and build quality). Of course applications are free to have multiple instances that have the same transform (of different grouped objects), but we can definitively state that the application _not_ expressing common transformations this way is only worse for the device to achieve equivalent performance (bookkeeping or searching for common transform matrices is required to recover this information). Thus we do not expect implementations to go out of their way to reverse-engineer information the application ought to already have -- somebody has to compute what transforms are common, so ANARI assumes the application would optimize for this in their ANARIWorld construction if it is important (some applications may simply not care, such as offline renderers).

So what about multi-level instancing? The current spec is only the minimally required parameters required by an implementation, but we have left the design open to extensions that permit `ANARIGroup` to contain an array of `ANARIInstance`, which expresses the nesting of an instance. No ANARI API changes are required for such an extension to be implemented. There is something to be said, though, about the performance of rendering and updating multi-level instanced scenes that is very non-trivial to answer -- multi-level instancing has cost (especially for ray tracers), and almost always has fixed depth limits (Embree and OptiX at least have this), while other APIs simply don't allow it at all (DXR + Vulkan RT) which forces the flattening operation to be opaque and thus unwieldy for applications to intuit its lifetime or overhead. This means that for what seems like a simple feature can actually add lots of entropy to performance intuition to what the cost of using such a feature entails. Again ANARI permits this as an exension, but we did not include it as core due to insufficient clarity that it is _better_ to be underneath the ANARI API instead of being the responsibility of the application's scene graph.

One idea we are wanting to explore is the idea of having transform _objects_, which would cut down on the number of updates required for multiple instances. Roughly speaking, instances could take a matrix directly (as they do now) _or_ take a handle to a transform, which could exist in more than one instance (set value on the handle once, inherit the value everywhere the transform object was set). This also gives us a chance to have different types of transforms too, such as blurred transforms or even stacked transforms. However, this design was only mentioned as a promising direction, but I am aware of no concrete progress on it yet.

In short, ANARI is trying to be as "thin" as possible between the application and the rendering engine by using a render graph object hierarchy that strives to avoid as much opaque flattening as possible. Given the diversity of implementation choices that rendering engines will make, I don't anticipate multi-level instancing being a required core feature any time soon. However, as soon as an ANARI device implementation successfully implements it as a vendor extension, we would very much consider including it as a core extension to give the feature a consistent design across devices.

Hope this helps!

Cheers,
Jeff

**Comment by pc-john** [comment]
Thanks Jeff. You wrote perfect summary, expressing the difference between rasterization and ray-tracing trade offs. What I see as the most important point for ray-tracing is this using of groups to express a common coordinate system.

You also mentioned Transform objects. I am a big fan of the idea of transform objects. I believe that they could bring great new possibilities. Let me mention some:

1. transformation is shared among attached instances - You already mentioned this. We change one Transform object and all attached instances will get new transformation. The benefit is clear: We might save many updates of transformation matrices.
2. we could have different types of Transforms, such as blurred, as you mentioned
3. hierarchical transformations - we might allow Transform objects to be attached to parent Transform objects and to have attached child Transform objects. You might not want to have this functionality directly in ANARI core, but I see number of benefits of such approach. There might be (a) further reduction of updates of transformation matrices and (b) because we have all the matrices passed through Anari API, we might make all matrix multiplications by ANARI device implementation, possibly using GPU to speed up the computation. It is not much for small scenes, but for millions of objects in non-flat graphs it might make a difference.

When considering the benefits of having Transform objects, we might easily spot that each Transform object serves to provide one local coordinate system that is shared by all attached Instances. In other words, all attached `Instances` share common coordinate system. I might not see everything, but it might void the benefits of "(geometry + material) --> group --> instance formulation". Without transform objects, I see clearly the intention of (geometry+material)->group->instance and I understand that it was chosen even although there is an overhead of 6 objects allocation for single thing rendered in the scene, e.g. we need 1xInstance, 1xGroup, 1x Array, 1xShape, 1xMaterial and 1xGeometry. For 1000 objects having the same Geometry but different placements and different materials, it is 1000xInstance + 1000xGroup + 1000xArray + 1000xShape + 1000xMaterial + 1xGeometry (5001 handles). I hope, I understood ANARI right. Otherwise, please, correct me.

However, with `Transform` objects, we might have more efficient design unless I am missing something. If we modify Instance object to contain only reference to Geometry, Material and Transform, we would allocate only 1000 Instances + 1000 Transforms + 1000 Material + 1 Geometry (3001 handles). Each Instance would express its "membership" in particular coordinate system by its attachment to particular Transform object. And each Instance would express its membership in particular group of objects having the same material by its attachment to particular Material object. This way, it would be easy for ray-tracing systems to optimize their BVH rebuilds based on Instance-Transform relation and it would be easy for rasterization systems to optimize for pipeline switches (and pipeline state changes in general) based on Instance-Material relation. Still other system might use even hybrid approaches, so, in some sense, we embrace all of them in a kind of easy to understand way.

Overall, introducing Transform objects and modifying Instance would include these benefits:

-  more efficient Instance position updates
-  allows for further optimizations like GPU evaluation of large Transform graphs
-  possibly lower number of allocated handles (possibly associated additional memory consumption and allocation overhead)
-  it would more clearly express the relation between Instance, its coordinate system given by Transform and its rendering state given by Material

One more thought: The current design leads to grouping all scene objects that share the same coordinate system bellow single Instance object. These scene objects might be separated objects and not intended to be considered one instance. But for the sake of ray-tracing performance, we would place them bellow single Instance. I feel like we are misusing the word instance here, using it in the place of coordinate system or something like that.

My goal here is definitely not to criticize. Rather, my goal is ANARI quality before it is released as the final spec. Therefore these thoughts about the efficiency of the scene graph design. Many of my thoughts might turn not relevant, but, maybe, some will contribute to ANARI quality. I am not paid for quality improvements of ANARI, so I cannot do too much at the moment and put my hands fully in, but I hope that ANARI will become important in the future.
